src/main/java/org/deepsymmetry/beatlink/AnnouncementSocketConnection.java

```diff
@@ -0,0 +1,294 @@
+                            }
+                            announcementSocket.get().receive(packet);
+                            received = !ignoredAddresses.contains(packet.getAddress());
+                        } catch (SocketTimeoutException ste) {
+                            received = false;
+                        } catch (IOException e) {
+                            // Don't log a warning if the exception was due to the socket closing at shutdown.
+                            if (isRunning()) {
+                                // We did not expect to have a problem; log a warning and shut down.
+                                logger.warn("Problem reading from DeviceAnnouncement socket, stopping", e);
+                                stop();
+                            }
+                            received = false;
+                        }
+                        try {
+                            if (received) {
+                                final Util.PacketType kind = Util.validateHeader(packet, ANNOUNCEMENT_PORT);
+                                if (kind == Util.PacketType.DEVICE_KEEP_ALIVE) {
+                                    // Looks like the kind of packet we need
+                                    if (packet.getLength() < 54) {
+                                        logger.warn("Ignoring too-short " + kind.name + " packet; expected 54 bytes, but only got " +
+                                                packet.getLength() + ".");
+                                    } else {
+                                        if (packet.getLength() > 54) {
+                                            logger.warn("Processing too-long " + kind.name + " packet; expected 54 bytes, but got " +
+                                                    packet.getLength() + ".");
+                                        }
+                                        DeviceAnnouncement announcement = new DeviceAnnouncement(packet);
+
+                                        deliverDeviceAnnouncement(announcement);
+                                    }
+                                } else if (kind == Util.PacketType.DEVICE_HELLO) {
+                                    logger.debug("Received device hello packet.");
+                                } else if (kind != null) {
+                                    // TODO, probably should be a set of listeners to not couple VirtualCdj to this class
+                                    VirtualCdj.getInstance().handleSpecialAnnouncementPacket(kind, packet);
+                                }
+                            }
+                        } catch (Throwable t) {
+                            logger.warn("Problem processing DeviceAnnouncement packet", t);
+                        }
+                    }
+                }
+            }, "beat-link DeviceFinder receiver");
```

Comment:
This thread should be renamed to reflect the class it has been moved to.

Reply:
Done.